QueryableEntry.serializationService is null when MapReduce is run with Extractable entries

Hazelcast version is 3.6.3.

If we try to run MapReduce with Extractable entries, a `NullPointerException` is thrown with the stacktrace below:

```
Exception in thread "main" com.hazelcast.core.HazelcastException: java.util.concurrent.ExecutionException: java.lang.NullPointerException
	at com.hazelcast.map.impl.proxy.MapProxyImpl.aggregate(MapProxyImpl.java:763)
	at com.hazelcast.map.impl.proxy.MapProxyImpl.aggregate(MapProxyImpl.java:734)
	at FXPBFeesPOC.main(FXPBFeesPOC.java:163)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:144)
Caused by: java.util.concurrent.ExecutionException: java.lang.NullPointerException
	at com.hazelcast.mapreduce.impl.task.TrackableJobFuture.setResult(TrackableJobFuture.java:68)
	at com.hazelcast.mapreduce.impl.task.JobSupervisor.notifyRemoteException(JobSupervisor.java:156)
	at com.hazelcast.mapreduce.impl.MapReduceUtil.notifyRemoteException(MapReduceUtil.java:86)
	at com.hazelcast.mapreduce.impl.task.MapCombineTask.handleProcessorThrowable(MapCombineTask.java:293)
	at com.hazelcast.mapreduce.impl.task.MapCombineTask.access$500(MapCombineTask.java:68)
	at com.hazelcast.mapreduce.impl.task.MapCombineTask$PartitionBasedProcessor.processPartitions(MapCombineTask.java:358)
	at com.hazelcast.mapreduce.impl.task.MapCombineTask$PartitionBasedProcessor.run(MapCombineTask.java:333)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
	at com.hazelcast.util.executor.CompletableFutureTask.run(CompletableFutureTask.java:67)
	at com.hazelcast.util.executor.CachedExecutorServiceDelegate$Worker.run(CachedExecutorServiceDelegate.java:212)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
	at com.hazelcast.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:76)
	at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:92)
Caused by: java.lang.NullPointerException
	at com.hazelcast.query.impl.getters.Extractors.getTargetObject(Extractors.java:80)
	at com.hazelcast.query.impl.getters.Extractors.extract(Extractors.java:62)
	at com.hazelcast.query.impl.QueryableEntry.extractAttributeValueFromTargetObject(QueryableEntry.java:144)
	at com.hazelcast.query.impl.QueryableEntry.extractAttributeValue(QueryableEntry.java:82)
	at com.hazelcast.query.impl.QueryableEntry.getAttributeValue(QueryableEntry.java:48)
	at com.hazelcast.query.impl.predicates.AbstractPredicate.readAttributeValue(AbstractPredicate.java:130)
	at com.hazelcast.query.impl.predicates.AbstractPredicate.apply(AbstractPredicate.java:55)
	at com.hazelcast.mapreduce.aggregation.impl.PredicateSupplier.apply(PredicateSupplier.java:58)
	at com.hazelcast.mapreduce.aggregation.impl.SupplierConsumingMapper.map(SupplierConsumingMapper.java:60)
	at com.hazelcast.mapreduce.impl.task.KeyValueSourceMappingPhase.executeMappingPhase(KeyValueSourceMappingPhase.java:49)
	at com.hazelcast.mapreduce.impl.task.MapCombineTask.processMapping(MapCombineTask.java:135)
	at com.hazelcast.mapreduce.impl.task.MapCombineTask.processPartitionMapping(MapCombineTask.java:304)
	at com.hazelcast.mapreduce.impl.task.MapCombineTask.access$700(MapCombineTask.java:68)
	at com.hazelcast.mapreduce.impl.task.MapCombineTask$PartitionBasedProcessor.processPartitions(MapCombineTask.java:356)
	at com.hazelcast.mapreduce.impl.task.MapCombineTask$PartitionBasedProcessor.run(MapCombineTask.java:333)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
	at com.hazelcast.util.executor.CompletableFutureTask.run(CompletableFutureTask.java:67)
	at com.hazelcast.util.executor.CachedExecutorServiceDelegate$Worker.run(CachedExecutorServiceDelegate.java:212)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
	at java.lang.Thread.run(Thread.java:745)
	at com.hazelcast.util.executor.HazelcastManagedThread.executeRun(HazelcastManagedThread.java:76)
	at com.hazelcast.util.executor.HazelcastManagedThread.run(HazelcastManagedThread.java:92)
	at ------ End remote and begin local stack-trace ------.(Unknown Source)
	at ------ Operation failed on node: Address[192.168.2.74]:5721 ------.notifyRemoteException(JobSupervisor.java:154)
	at com.hazelcast.mapreduce.impl.task.JobSupervisor.notifyRemoteException(JobSupervisor.java:154)
	... 13 more
```

--- a/hazelcast/src/main/java/com/hazelcast/mapreduce/aggregation/impl/DistinctValuesAggregation.java
+++ b/hazelcast/src/main/java/com/hazelcast/mapreduce/aggregation/impl/DistinctValuesAggregation.java
@@ -24,6 +24,7 @@ import com.hazelcast.mapreduce.Mapper;
 import com.hazelcast.mapreduce.Reducer;
 import com.hazelcast.mapreduce.ReducerFactory;
 import com.hazelcast.mapreduce.aggregation.Supplier;
+import com.hazelcast.mapreduce.impl.task.DefaultContext;
 import com.hazelcast.nio.ObjectDataInput;
 import com.hazelcast.nio.ObjectDataOutput;
 import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
@@ -199,6 +200,7 @@ public class DistinctValuesAggregation<Key, Value, DistinctType>
             int mappingKey = key();
             entry.setKey(key);
             entry.setValue(value);
+            entry.setSerializationService(((DefaultContext) context).getSerializationService());
             DistinctType valueOut = supplier.apply(entry);
             if (valueOut != null) {
                 context.emit(mappingKey, valueOut);

--- a/hazelcast/src/main/java/com/hazelcast/mapreduce/aggregation/impl/SimpleEntry.java
+++ b/hazelcast/src/main/java/com/hazelcast/mapreduce/aggregation/impl/SimpleEntry.java
@@ -17,6 +17,7 @@
 package com.hazelcast.mapreduce.aggregation.impl;
 
 import com.hazelcast.config.MapAttributeConfig;
+import com.hazelcast.internal.serialization.InternalSerializationService;
 import com.hazelcast.nio.serialization.Data;
 import com.hazelcast.query.impl.QueryableEntry;
 import com.hazelcast.query.impl.getters.Extractors;
@@ -76,6 +77,10 @@ final class SimpleEntry<K, V>
 
     void setKey(K key) {
         this.key = key;
+    }
+
+    void setSerializationService(InternalSerializationService serializationService) {
+        this.serializationService = serializationService;
     }
 
 }

--- a/hazelcast/src/main/java/com/hazelcast/mapreduce/aggregation/impl/SupplierConsumingMapper.java
+++ b/hazelcast/src/main/java/com/hazelcast/mapreduce/aggregation/impl/SupplierConsumingMapper.java
@@ -19,6 +19,7 @@ package com.hazelcast.mapreduce.aggregation.impl;
 import com.hazelcast.mapreduce.Context;
 import com.hazelcast.mapreduce.Mapper;
 import com.hazelcast.mapreduce.aggregation.Supplier;
+import com.hazelcast.mapreduce.impl.task.DefaultContext;
 import com.hazelcast.nio.ObjectDataInput;
 import com.hazelcast.nio.ObjectDataOutput;
 import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
@@ -52,6 +53,7 @@ class SupplierConsumingMapper<Key, ValueIn, ValueOut>
     public void map(Key key, ValueIn value, Context<Key, ValueOut> context) {
         entry.setKey(key);
         entry.setValue(value);
+        entry.setSerializationService(((DefaultContext) context).getSerializationService());
         ValueOut valueOut = supplier.apply(entry);
         if (valueOut != null) {
             context.emit(key, valueOut);

--- a/hazelcast/src/main/java/com/hazelcast/mapreduce/impl/task/DefaultContext.java
+++ b/hazelcast/src/main/java/com/hazelcast/mapreduce/impl/task/DefaultContext.java
@@ -17,6 +17,7 @@
 package com.hazelcast.mapreduce.impl.task;
 
 import com.hazelcast.core.IFunction;
+import com.hazelcast.internal.serialization.InternalSerializationService;
 import com.hazelcast.mapreduce.Combiner;
 import com.hazelcast.mapreduce.CombinerFactory;
 import com.hazelcast.mapreduce.Context;
@@ -32,6 +33,7 @@ import java.util.Map;
 import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
 
 import static com.hazelcast.util.ConcurrentReferenceHashMap.ReferenceType.STRONG;
+
 /**
  * This is the internal default implementation of a map reduce context mappers emit values to. It controls the emitted
  * values to be combined using either the set {@link com.hazelcast.mapreduce.Combiner} or by utilizing the internal
@@ -66,6 +68,8 @@ public class DefaultContext<KeyIn, ValueIn>
     private volatile int collected;
 
     private volatile int partitionId;
+
+    private volatile InternalSerializationService serializationService;
 
     protected DefaultContext(CombinerFactory<KeyIn, ValueIn, ?> combinerFactory, MapCombineTask mapCombineTask) {
         this.mapCombineTask = mapCombineTask;
@@ -111,7 +115,15 @@ public class DefaultContext<KeyIn, ValueIn>
     }
 
     public Combiner<ValueIn, ?> getOrCreateCombiner(KeyIn key) {
-        return combiners.applyIfAbsent(key , combinerFunction);
+        return combiners.applyIfAbsent(key, combinerFunction);
+    }
+
+    public void setSerializationService(InternalSerializationService serializationService) {
+        this.serializationService = serializationService;
+    }
+
+    public InternalSerializationService getSerializationService() {
+        return serializationService;
     }
 
     /**

--- a/hazelcast/src/main/java/com/hazelcast/mapreduce/impl/task/MapCombineTask.java
+++ b/hazelcast/src/main/java/com/hazelcast/mapreduce/impl/task/MapCombineTask.java
@@ -16,6 +16,7 @@
 
 package com.hazelcast.mapreduce.impl.task;
 
+import com.hazelcast.internal.serialization.InternalSerializationService;
 import com.hazelcast.mapreduce.KeyValueSource;
 import com.hazelcast.mapreduce.LifecycleMapper;
 import com.hazelcast.mapreduce.Mapper;
@@ -35,6 +36,7 @@ import com.hazelcast.mapreduce.impl.operation.RequestPartitionResult;
 import com.hazelcast.nio.Address;
 import com.hazelcast.spi.NodeEngine;
 import com.hazelcast.spi.partition.IPartitionService;
+import com.hazelcast.spi.serialization.SerializationService;
 import com.hazelcast.util.ExceptionUtil;
 
 import java.util.Collection;
@@ -74,6 +76,7 @@ public class MapCombineTask<KeyIn, ValueIn, KeyOut, ValueOut, Chunk> {
     private final KeyValueSource<KeyIn, ValueIn> keyValueSource;
     private final MapReduceService mapReduceService;
     private final IPartitionService partitionService;
+    private final SerializationService serializationService;
     private final JobSupervisor supervisor;
     private final NodeEngine nodeEngine;
     private final String name;
@@ -91,6 +94,7 @@ public class MapCombineTask<KeyIn, ValueIn, KeyOut, ValueOut, Chunk> {
         this.chunkSize = configuration.getChunkSize();
         this.nodeEngine = configuration.getNodeEngine();
         this.partitionService = nodeEngine.getPartitionService();
+        this.serializationService = nodeEngine.getSerializationService();
         this.mapReduceService = supervisor.getMapReduceService();
         this.keyValueSource = configuration.getKeyValueSource();
     }
@@ -126,6 +130,7 @@ public class MapCombineTask<KeyIn, ValueIn, KeyOut, ValueOut, Chunk> {
             throws Exception {
 
         context.setPartitionId(partitionId);
+        context.setSerializationService((InternalSerializationService) serializationService);
 
         if (mapper instanceof LifecycleMapper) {
             ((LifecycleMapper) mapper).initialize(context);

--- a/hazelcast/src/test/java/com/hazelcast/mapreduce/aggregation/MapCountAggregationPortableTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/mapreduce/aggregation/MapCountAggregationPortableTest.java
@@ -1,0 +1,90 @@
+package com.hazelcast.mapreduce.aggregation;
+
+import com.hazelcast.config.Config;
+import com.hazelcast.core.Hazelcast;
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.core.IMap;
+import com.hazelcast.nio.serialization.Portable;
+import com.hazelcast.nio.serialization.PortableFactory;
+import com.hazelcast.nio.serialization.PortableReader;
+import com.hazelcast.nio.serialization.PortableWriter;
+import com.hazelcast.query.Predicates;
+import com.hazelcast.test.HazelcastSerialClassRunner;
+import com.hazelcast.test.HazelcastTestSupport;
+import com.hazelcast.test.annotation.QuickTest;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+
+import java.io.IOException;
+
+import static org.junit.Assert.assertEquals;
+
+@RunWith(HazelcastSerialClassRunner.class)
+@Category({QuickTest.class})
+public class MapCountAggregationPortableTest extends HazelcastTestSupport {
+
+    private HazelcastInstance hazelcastInstance;
+
+    @Before
+    public void before() {
+        Config config = new Config();
+        config.getSerializationConfig().addPortableFactory(1, new FooPortableFactory());
+        this.hazelcastInstance = Hazelcast.newHazelcastInstance(config);
+    }
+
+    @After
+    public void after() {
+        this.hazelcastInstance.shutdown();
+    }
+
+    @Test
+    public void testCountAggregationWithPortable() {
+        // GIVEN
+        IMap<Integer, Foo> fooMap = hazelcastInstance.getMap("fooMap");
+        Foo value = new Foo();
+        value.text = "foo";
+        fooMap.put(1, value);
+
+        // WHEN
+        Supplier<Integer, Foo, Integer> supplier = Supplier.fromPredicate(Predicates.equal("text", "foo"));
+        Aggregation<Integer, Integer, Long> counter = Aggregations.count();
+        long count = fooMap.aggregate(supplier, counter);
+
+        // THEN
+        assertEquals(1L, count);
+    }
+
+    private static class FooPortableFactory implements PortableFactory {
+        public Portable create(int classId) {
+            if (Foo.ID == classId) {
+                return new Foo();
+            } else {
+                return null;
+            }
+        }
+    }
+
+    private static class Foo implements Portable {
+        public static final int ID = 1;
+        String text;
+
+        public int getFactoryId() {
+            return 1;
+        }
+
+        public int getClassId() {
+            return ID;
+        }
+
+        public void writePortable(PortableWriter writer) throws IOException {
+            writer.writeUTF("text", text);
+        }
+
+        public void readPortable(PortableReader reader) throws IOException {
+            text = reader.readUTF("text");
+        }
+    }
+}

